### PR TITLE
no-restricted-imports: deprecate mui/styled in our frontend code in f…

### PR DESCRIPTION
**Pull request by wimagguc**
### Public-Facing Changes

A `@foxglove/frontend-mui-recommended` config to prevent `mui/styles` in new components. It uses the `@foxglove/no-restricted-imports` rule to achieve this, which is the eslint built-in rule re-exported for safety. This in an effort to avoid accidental overrides.

### CODE REVIEW

**Comment by defunctzombie** [comment]
This should go into `foxglove/console` rather than be in our common react. I'd accept an alternative here which puts these into a `configs/frontend-ui-recommended` file or some such.

Lets see what others say about how to organize this.

**Comment by amacneil** [comment]
It was my suggestion to move here - seemed like something relevant to all our frontend projects. I don't feel strongly about it though.

One gotcha I will note is that if you set `no-restricted-imports` in this plugin, and another project imports this plugin but then tries to add additional restricted imports, that will overwrite (not append to) the defaults in this shared config. We solved that in the past by creating a new rule in https://github.com/foxglove/eslint-plugin/tree/main/rules rather than using a rule that might likely get overridden.

**Comment by defunctzombie** [comment]
I am fine having it in this repo but we should put it in a separate file you can include like `frontend-mui-recommended` and I agree with @amacneil comment that we should make it so changing `no-restricted-imports` in the repo does not remove this.

**Comment by wimagguc** [comment]
> I am fine having it in this repo but we should put it in a separate file you can include like `frontend-mui-recommended` and I agree with @amacneil comment that we should make it so changing `no-restricted-imports` in the repo does not remove this.

Done and done, updated the description accordingly. Thanks!

**Comment by amacneil** [comment]
@defunctzombie what's the logic for having it be a separate config file - isn't this whole repo "recommended" config? Separate projects can disable the config if they do not want to use it, but wouldn't it be better to keep all our best practices in one default config?

**Review by defunctzombie — APPROVED** [review]
You could drop the "frontend" from the name and filename `mui-recommended` is pretty good on its own.

**Comment by defunctzombie** [comment]
@amacneil same logic we have separate configs for typescript, react, base. Gives you a choice of what to include. I would be fine throwing this into the react one fwiw. I didn't know we had one for that already.

**Comment by amacneil** [comment]
Yeah, maybe react is a good place for it.

From memory we created that separation because each of those libraries had actual dependencies (e.g. typescript rules cannot be enabled on a JS project without typescript installed, I think similar for react). So that limitation doesn't really apply here. But putting it in the react config seems fine.

**Comment by wimagguc** [comment]
> Yeah, maybe react is a good place for it.
> 
> From memory we created that separation because each of those libraries had actual dependencies (e.g. typescript rules cannot be enabled on a JS project without typescript installed, I think similar for react). So that limitation doesn't really apply here. But putting it in the react config seems fine.

...right, so the issue with moving this to `@foxglove/react` was that now we need to update studio's eslint rules as well, because it has its own mui-related rules in `no-restricted-imports` that comes with a list of exceptions. I guess the question is:
(a) Leave the rules in `react` and update Studio, or
(b) Revert the latest change, keep using `mui-recommended` and revisit merging this with `react` when studio and console are in the same place

File: configs/react.js [hunk 1]
```diff
@@ -16,5 +16,31 @@ module.exports = {
 
     // require rel="noopener" or rel="noreferrer" with target="_blank" for security
     "react/jsx-no-target-blank": ["error", { allowReferrer: true }],
+
+    // deprecated mui/styled in our frontend code in favor of tss-react/mui
+    "@foxglove/no-restricted-imports": [
```
Comment (amacneil) [inline configs/react.js:21]:
```suggestion
    "@foxglove/no-mui-styled": [
```

Probably better to give this a more specific name if the list of rules is specific to mui styled.

**Comment by amacneil** [comment]
"(a) Leave the rules in react and update Studio" sounds good.

We should try to be consistent with style and patterns across repos. You can update the list of exceptions in studio to ignore this new rule instead.

File: configs/react.js (hunk 1 above)
Comment (wimagguc) [reply configs/react.js:21]: That custom rule is a passthrough of `no-restricted-imports` in eslint's built-in rules; it only becomes mui-specific when it's used within the `@foxglove/react` config with the mui params — and the config already has "react" in its name that makes it specific.
I don't think we need to rename `@foxglove/no-restricted-imports` here. That name tells us exactly what it is, how to use it and what we can use it for.

**Pull request merged**